content/postgresql/postgresql-18/array-bytea-improvements.md

```diff
@@ -0,0 +1,196 @@
+
+## Binary Data (Bytea) Improvements
+
+PostgreSQL 18 introduces several improvements for working with binary data, including a new reverse function and enhanced casting capabilities.
+
+### New reverse() Function for Bytea
+
+The `reverse()` function allows you to reverse the byte order of binary data:
+
+```sql
+SELECT reverse('\x12345678'::bytea);
+-- Result: \x78563412
+```
+
+This is particularly useful for endianness conversion between big-endian and little-endian formats, cryptographic operations that require byte reversal, and data format transformations.
+
+## Integer-to-Bytea Casting
+
+PostgreSQL 18 introduces direct casting between integer types and bytea, representing integers as two's complement binary values with the most significant byte first (big-endian format).
+
+### Basic Integer Casting
+
+```sql
+SELECT 1234::bytea;
+-- Result: \x000004d2
+
+SELECT (-1234)::bytea;
+-- Result: \xfffffb2e
+```
+
+Different integer types produce different byte lengths - the output size is fixed based on the integer type, not the value:
+
+```sql
+SELECT 123::smallint::bytea;  -- Always 2 bytes: \x007b
+SELECT 123::integer::bytea;   -- Always 4 bytes: \x0000007b
+SELECT 123::bigint::bytea;    -- Always 8 bytes: \x000000000000007b
+```
+
+In previous versions, if you were to try to cast an integer directly to bytea, you would have to use a custom function or complex bit manipulation. Now, this is simplified with direct casting.
+
+Whether you cast `123` or `123456`, an `integer::bytea` conversion always produces exactly 4 bytes.
+
+You can also cast back from bytea to integers:
+
+```sql
+SELECT '\x007b'::bytea::smallint;
+-- Result: 123
+```
+
+This feature is useful for compact binary storage of numeric IDs, protocol implementations that require specific byte representations, and interfacing with external systems that expect binary integer formats.
+
+## MIN/MAX Aggregates for Arrays and Composite Types
+
+PostgreSQL 18 extends `MIN()` and `MAX()` aggregate functions to work with arrays and composite types, using their natural comparison operators.
+
+### Array Aggregates
+
+Arrays are compared element by element (lexicographically). This means `{1,3}` is considered less than `{2,1}` because the first element is smaller.
+
+Let's see how this works with arrays:
+
+```sql
+-- Sample table with arrays
+CREATE TABLE sales_data (
+    product VARCHAR(50),
+    monthly_sales INTEGER[]
+);
+
+INSERT INTO sales_data VALUES
+    ('Laptop', ARRAY[45, 52, 38]),
+    ('Mouse', ARRAY[67, 71, 58]),
+    ('Keyboard', ARRAY[23, 28, 15]);
+
+-- Find minimum and maximum sales arrays
+SELECT
+    MIN(monthly_sales) AS min_sales_pattern,
+    MAX(monthly_sales) AS max_sales_pattern
+FROM sales_data;
+```
+
+This query returns the lexicographically smallest and largest sales patterns across products based on the first element of each array.
```

Comment:
```suggestion
This query returns the `Keyboard` monthly_sales value for min and the `Mouse` value for max because it is evaluating based on the first element of each array.
```